Correction of Generic Contaminant Handling in Some Situations

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #10473 
 - While development team members were working on other defects, it was noticed that there were a couple of places in the code where the generic contaminant information was not being handled properly, i.e. it wasn't being updated properly.  This pull request fixes this issue and now all code should be handling generic contaminant updating properly.  As part of this fix, a thorough check of the entire code was made to ensure that both CO2 and generic contaminant code was present in all locations.  During that check, an unnecessary/misplaced line of code was found and eliminated as well.  While no documentation fixes were need as part of this work, a new unit test was created and an existing unit test was enhanced to test this correction.  No changes in the output were noted and the unit tests show that updating is being handled correctly now.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Fix for GenContam issues

There were a few places where GenContam was not implemented properly in the code.  After a thorough search of the code for both CO2 and GenContam, it was discovered that there were a couple of places where the code for these two entities were not similar (GenContam code that needed to be present was missing).  Also, an incorrect assignment of the simulation flag was also discovered.  This commit fixes those issues.

RKStrand committed:
Merge branch 'develop' into 10473CorrectCO2GenContamHandling

RKStrand committed:
Unit test for GenContam fix

Addition of a new unit test and enhancement of an existing one for this fix.

RKStrand committed:
Correction of unit test

To const or to constexpr, that is the question...

mjwitte commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -1810,12 +1811,14 @@ void ZoneEquipmentMixer::setOutletConditions(EnergyPlusData &state)
         sumEnthalpy += spaceOutletNode.Enthalpy * mixerSpace.fraction;
         sumHumRat += spaceOutletNode.HumRat * mixerSpace.fraction;
         sumCO2 += spaceOutletNode.CO2 * mixerSpace.fraction;
+        sumGenContam += spaceOutletNode.GenContam * mixerSpace.fraction;
```
Most (all?) other places that work with CO2 and GenContam have ifs like `if (state.dataContaminantBalance->Contaminant.GenericContamSimulation) {`. Should there be ifs here?

RKStrand replied:
@mjwitte hmmm…that’s a good point. I will look at implementing such if checks when I get back to this next week on Tuesday.

RKStrand replied:
@mjwitte Ok, I've added the IF checks as per your suggestion and just now made a commit (as well as merged in develop to get everything back up to speed).  Anticipating that this should come back all green as before.

mjwitte commented on src/EnergyPlus/VariableSpeedCoils.cc:
```diff
@@ -8337,6 +8337,10 @@ namespace VariableSpeedCoils {
             state.dataLoopNodes->Node(AirOutletNode).CO2 = state.dataLoopNodes->Node(AirInletNode).CO2;
         }
 
+        if (state.dataContaminantBalance->Contaminant.GenericContamSimulation) {
+            state.dataLoopNodes->Node(AirOutletNode).GenContam = state.dataLoopNodes->Node(AirInletNode).GenContam;
+        }
```
@Myoldmopar There are ifs like this scattered throughout the code. For simple assignments like this are these ifs saving any time or are they just clutter?

rraustad replied:
I agree these checks aren't needed (just pass the 0). Also thinking of a safeAirNodeCopy function passing MassFlowRate, Quality, Press, MassFlowRateMin, MassFlowRateMax, MassFlowRateMinAvail and MassFlowRateMaxAvail, and CO2 and GenContam. A lot of these child update functions would reduce to:

        safeAirNodeCopy(state, outletnode, inletnode);
        state.dataLoopNodes->Node(AirOutletNode).Temp = varSpeedCoil.OutletAirDBTemp;
        state.dataLoopNodes->Node(AirOutletNode).HumRat = varSpeedCoil.OutletAirHumRat;
        state.dataLoopNodes->Node(AirOutletNode).Enthalpy = varSpeedCoil.OutletAirEnthalpy;

RKStrand committed:
Things are getting a bit IFfy

To conform to the pattern established in other parts of the code, IF statements were needed in the setOutletConditions routine of the ZoneEquipmentMixer.  This also then required the slight modification of a unit test.  This was in response to a comment from @mjwitte.

RKStrand committed:
Merge branch 'develop' into 10473CorrectCO2GenContamHandling

mjwitte reviewed (approved):
Code changes look good. Unit tests pass. CI is clean and green. Merging. Thanks @RKStrand !
